vnext/JSI/Shared/ChakraRuntime.cpp

```diff
@@ -607,284 +616,202 @@ facebook::jsi::Runtime::ScopeState *ChakraRuntime::pushScope() {
-
+      break;
+    }
     case JsNumber: {
-      double doubleValue;
-      JsNumberToDouble(value, &doubleValue);
-      facebook::jsi::Value val(doubleValue);
-      return val;
+      double number;
+      ThrowUponJsError(JsNumberToDouble(ref, &number), "JsNumberToDouble");
+      return facebook::jsi::Value(number);
+      break;
     }
-
     case JsString: {
-      std::string utf8str = JSStringToSTLString(value);
-      return facebook::jsi::String::createFromUtf8(
-          *const_cast<facebook::jsi::Runtime *>(
-              reinterpret_cast<const facebook::jsi::Runtime *>(this)),
-          reinterpret_cast<const uint8_t *>(utf8str.c_str()),
-          utf8str.size());
+      return facebook::jsi::Value(
+          *this, MakePointer<facebook::jsi::String>(std::move(ref)));
       break;
     }
-
     case JsBoolean: {
-      bool boolValue;
-      JsBooleanToBool(value, &boolValue);
-      facebook::jsi::Value val(boolValue);
-      return val;
+      bool b;
+      ThrowUponJsError(JsBooleanToBool(ref, &b), "JsBooleanToBool");
+      return facebook::jsi::Value(b);
+      break;
     }
-
-    case JsObject:
-    case JsFunction:
-    case JsArray: {
-      return facebook::jsi::Value(createObject(value));
+    case JsSymbol: {
+      return facebook::jsi::Value(
+          *this, MakePointer<facebook::jsi::Symbol>(std::move(ref)));
       break;
     }
-
+    case JsObject:
+    case JsFunction:
     case JsError:
-    case JsSymbol:
+    case JsArray:
     case JsArrayBuffer:
     case JsTypedArray:
-    case JsDataView:
-    default:
+    case JsDataView: {
+      return facebook::jsi::Value(
+          *this, MakePointer<facebook::jsi::Object>(std::move(ref)));
+      break;
+    }
+    default: {
+      // Control flow should never reach here.
       std::terminate();
       break;
+    }
   }
-}
-
-JsValueRef ChakraRuntime::valueRef(const facebook::jsi::Value &valueIn) {
-  if (valueIn.isUndefined()) {
-    JsValueRef value;
-    JsGetUndefinedValue(&value);
-    return value;
-  } else if (valueIn.isNull()) {
-    JsValueRef value;
-    JsGetNullValue(&value);
-    return value;
-  } else if (valueIn.isBool()) {
-    JsValueRef value;
-    JsBoolToBoolean(valueIn.getBool(), &value);
-    return value;
-  } else if (valueIn.isNumber()) {
-    JsValueRef value;
-    JsDoubleToNumber(valueIn.getNumber(), &value);
-    return value;
-  } else if (valueIn.isString()) {
-    return stringRef(valueIn.getString(*this));
-  } else if (valueIn.isObject()) {
-    return objectRef(valueIn.getObject(*this));
-  } else {
-    // What are you?
-    abort();
-  }
-}
 
-ChakraRuntime::ChakraPropertyIdValue::~ChakraPropertyIdValue() {
-  JsRelease(m_propId, nullptr);
+  // Control flow should never reach here.
+  std::terminate();
 }
 
-void ChakraRuntime::ChakraPropertyIdValue::invalidate() {
-  delete this;
-}
+ChakraObjectRef ChakraRuntime::ToChakraObjectRef(
+    const facebook::jsi::Value &value) {
+  if (value.isUndefined()) {
```

Comment:
>if (value.isUndefined()) { [](start = 2, length = 26)

(Nit) There is some duplication here for the various JsValueRef types; see if there's an opportunity to refactor